pkg/ttnpb/udp/packet_data.go

```diff
@@ -53,6 +53,23 @@ type RxPacket struct {
 	Aesk  uint         `json:"aesk"`            // AES key index used for encrypting fine timestamps (unsigned integer)
 }
 
+// reduceCodingRate reduces the coding rate fraction returned by the packet forwarder.
+// The packet forwarder is concentrator centric - and the concentrator expresses coding
+// rates as a fraction of form `4/(4 + x)`. But the RP002-1.0.x series of documents uses
+// the irreducible form of the fraction.
+func reduceCodingRate(cr string) string {
+	switch cr {
+	case "4/6":
+		return "2/3"
+	case "4/8":
+		return "1/2"
+	case "2/6":
+		return "1/3"
+	default:
+		return cr
+	}
+}
```

Comment:
Isn't this `4/(4 + X)` notation specific to LoRa modulation?

The coding rate has the useful bits on the numerator and the total bits on the denominator. So `4/(4 + X)` means that the redundant bits are `X`. My interpretation is that in LoRa modulation, the useful bits are always a multiple of 4, and so is the nominator. For example, the coding rate is `4/8` and there are 9 useful bits to send. `4/8` would make it explicit that `ceil(9/4)*8 = 24` bits are sent, while with `1/2` it would be `9*2 = 18` bits. But I might be wrong here.

To this function name; this is very generic and seems to apply only to LR-FHSS.

Reply:
I've added some new details in [this](https://github.com/TheThingsIndustries/lorawan-stack/issues/3400#issuecomment-1227693736) issue, given that the packet forwarder is not public yet. 

TLDR is that I think you are right, this is an packet forwarder quirk related to how data rates are rendered, and seems to apply only to LR-FHSS.